File: src/commands/dev/tag/create.rs
//! Git Tag 创建命令
//!
//! 提供创建和推送 Git tag 的功能。

use crate::git::{GitCommand, GitTag};
use crate::{log_break, log_error, log_info, log_success, log_warning};
use color_eyre::{eyre::WrapErr, Result};

/// Git Tag 创建命令
pub struct TagCreateCommand {
    tag: String,
    commit_sha: Option<String>,
    ci: bool,
}

impl TagCreateCommand {
    /// 创建新的 Tag 创建命令
    pub fn new(tag: String, commit_sha: Option<String>, ci: bool) -> Self {
        Self {
            tag,
            commit_sha,
            ci,
        }
    }

    /// 创建并推送 tag
    pub fn create(&self) -> Result<()> {
        log_break!('=');
        log_info!("创建 Git Tag");
        log_break!('=');
        log_break!();

        let tag_name = &self.tag;
        log_info!("Tag 名称: {}", tag_name);

        // 检查 tag 是否已存在
        let (exists_local, exists_remote) = GitTag::is_tag_exists(tag_name)?;

        if exists_local || exists_remote {
            log_warning!("Tag 已存在");
            log_info!("   本地存在: {}", exists_local);
            log_info!("   远程存在: {}", exists_remote);

            // 获取现有 tag 的 commit SHA
            let existing_tag_info = GitTag::get_tag_info(tag_name)?;
            let current_head = GitCommand::new(["rev-parse", "HEAD"]).read().unwrap_or_default();
            let target_sha = self.commit_sha.as_deref().unwrap_or(&current_head);

            log_info!("   现有 tag commit: {}", existing_tag_info.commit_hash);
            log_info!("   目标 commit: {}", target_sha);

            if existing_tag_info.commit_hash == target_sha {
                log_success!("Tag 已存在且指向正确的 commit");
                if self.ci {
                    self.output_ci_result(true, tag_name)?;
                }
                return Ok(());
            } else {
                log_warning!("Tag 已存在但指向不同的 commit");
                log_warning!("   删除现有 tag 并重新创建...");

                if exists_local {
                    GitTag::delete_local(tag_name)?;
                    log_success!("已删除本地 tag");
                }
                if exists_remote {
                    GitTag::delete_remote(tag_name)?;
                    log_success!("已删除远程 tag");
                }
            }
        }

        // 创建 tag
        log_break!();
        log_info!("创建 tag...");
        GitTag::create(tag_name, self.commit_sha.as_deref())?;
        log_success!("Tag 创建成功");

        // 推送 tag
        log_break!();
        log_info!("推送 tag 到远程...");
        match GitTag::push(tag_name) {
            Ok(()) => {
                log_success!("Tag 推送成功");
            }
            Err(e) => {
                // 检查 tag 是否已存在于远程且指向正确的 commit
                if let Ok(remote_tags) = GitTag::list_remote_tags() {
                    if remote_tags.contains(&tag_name.to_string()) {
                        let remote_tag_info = GitTag::get_tag_info(tag_name)?;
                        let current_head =
                            GitCommand::new(["rev-parse", "HEAD"]).read().unwrap_or_default();
                        let target_sha =
                            self.commit_sha.as_deref().unwrap_or(current_head.as_str());

                        if remote_tag_info.commit_hash == target_sha {
                            log_success!("Tag 已存在于远程且指向正确的 commit");
                            if self.ci {
                                self.output_ci_result(true, tag_name)?;
                            }
                            return Ok(());
                        } else {
                            log_error!("Tag 存在于远程但指向不同的 commit");
                            log_error!("   远程 tag commit: {}", remote_tag_info.commit_hash);
                            log_error!("   目标 commit: {}", target_sha);
                            return Err(e);
                        }
                    }
                }
                return Err(e);
            }
        }

        log_break!();
        log_success!("Tag 创建和推送完成");
        log_info!("   Tag: {}", tag_name);
        if let Some(ref sha) = self.commit_sha {
            log_info!("   Commit: {}", sha);
        } else {
            let current_sha = GitCommand::new(["rev-parse", "HEAD"]).read()?;
            log_info!("   Commit: {}", current_sha);
        }
        log_break!();

        // CI 模式：输出到 GITHUB_OUTPUT
        if self.ci {
            self.output_ci_result(true, tag_name)?;
        }

        Ok(())
    }

    /// 输出 CI 模式结果到 GITHUB_OUTPUT
    fn output_ci_result(&self, success: bool, tag_name: &str) -> Result<()> {
        use std::fs::OpenOptions;
        use std::io::Write;

        if let Ok(output_file) = std::env::var("GITHUB_OUTPUT") {
            let mut file = OpenOptions::new()
                .append(true)
                .create(true)
                .open(&output_file)
                .wrap_err_with(|| format!("Failed to open GITHUB_OUTPUT: {}", output_file))?;

            writeln!(file, "tag={}", tag_name).wrap_err("Failed to write tag")?;
            writeln!(file, "tag_created={}", success).wrap_err("Failed to write tag_created")?;
        }

        Ok(())
    }
}

File: scripts/dev/py/tag/create.py
"""Git Tag 创建实现"""
import os
import sys
from pathlib import Path
from typing import Optional

# 添加父目录到路径
sys.path.insert(0, str(Path(__file__).parent.parent))
from utils.logger import log_info, log_success, log_warning, log_error, log_break
from utils.git import (
    tag_exists, create_tag, delete_local_tag, delete_remote_tag,
    push_tag, get_last_commit_sha, run_git_command
)


def _get_tag_info(tag_name: str) -> Optional[str]:
    """获取 tag 指向的 commit SHA"""
    try:
        result = run_git_command(['rev-list', '-n', '1', tag_name], check=False)
        if result.returncode == 0:
            return result.stdout.strip()
    except Exception:
        pass
    return None


def _output_ci_result(created: bool, tag_name: str) -> None:
    """输出 CI 结果到 GITHUB_OUTPUT"""
    output_file = os.environ.get('GITHUB_OUTPUT')
    if output_file:
        try:
            with open(output_file, 'a') as f:
                f.write(f'tag_created={str(created).lower()}\n')
                f.write(f'tag_name={tag_name}\n')
        except IOError as e:
            log_error(f"Failed to write to GITHUB_OUTPUT: {e}")


def create(args) -> None:
    """创建并推送 tag"""
    log_break('=')
    log_info("创建 Git Tag")
    log_break('=')
    log_break()

    tag_name = args.tag
    log_info(f"Tag 名称: {tag_name}")

    # 检查 tag 是否已存在
    exists_local, exists_remote = tag_exists(tag_name)

    if exists_local or exists_remote:
        log_warning("Tag 已存在")
        log_info(f"   本地存在: {exists_local}")
        log_info(f"   远程存在: {exists_remote}")

        # 获取现有 tag 的 commit SHA
        existing_tag_sha = _get_tag_info(tag_name)
        current_head = get_last_commit_sha() or ""
        target_sha = args.commit or current_head

        if existing_tag_sha:
            log_info(f"   现有 tag commit: {existing_tag_sha}")
        log_info(f"   目标 commit: {target_sha}")

        if existing_tag_sha == target_sha:
            log_success("Tag 已存在且指向正确的 commit")
            if args.ci:
                _output_ci_result(False, tag_name)
            return
        else:
            log_warning("Tag 已存在但指向不同的 commit")
            log_warning("   删除现有 tag 并重新创建...")

            if exists_local:
                delete_local_tag(tag_name)
                log_success("已删除本地 tag")
            if exists_remote:
                delete_remote_tag(tag_name)
                log_success("已删除远程 tag")

    # 创建 tag
    log_break()
    log_info("创建 tag...")
    create_tag(tag_name, args.commit)
    log_success("Tag 创建成功")

    # 推送 tag
    log_break()
    log_info("推送 tag 到远程...")
    try:
        push_tag(tag_name)
        log_success("Tag 推送成功")
    except Exception as e:
        log_warning(f"Tag 推送失败（可能已存在）: {e}")
        # 验证远程 tag 是否指向正确的 commit
        try:
            result = run_git_command(['ls-remote', '--tags', 'origin', tag_name], check=False)
            if tag_name in result.stdout:
                remote_tag_sha = result.stdout.split()[0]
                if remote_tag_sha == target_sha:
                    log_success("Tag 已存在于远程且指向正确的 commit")
                else:
                    log_error("Tag 存在于远程但指向不同的 commit")
                    log_error(f"   远程 tag commit: {remote_tag_sha}")
                    log_error(f"   目标 commit: {target_sha}")
                    sys.exit(1)
        except Exception:
            pass

    log_success("Tag 创建和推送完成")
    log_info(f"   Tag: {tag_name}")
    if args.commit:
        log_info(f"   Commit: {args.commit}")
    else:
        current_sha = get_last_commit_sha()
        if current_sha:
            log_info(f"   Commit: {current_sha}")

    if args.ci:
        _output_ci_result(True, tag_name)


def main():
    """CLI 入口（可以直接运行此脚本）"""
    import argparse

    parser = argparse.ArgumentParser(
        prog='create',
        description='Create and push Git tag'
    )
    parser.add_argument('--tag', required=True, help='Tag name')
    parser.add_argument('--commit', help='Commit SHA (default: HEAD)')
    parser.add_argument('--ci', action='store_true', help='CI mode')

    args = parser.parse_args()
    create(args)


if __name__ == '__main__':
    main()

